spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -18,79 +18,93 @@
 import org.springframework.cloud.gcp.pubsub.support.SubscriberFactory;
 import org.springframework.cloud.gcp.pubsub.support.converter.PubSubMessageConverter;
-import org.springframework.cloud.gcp.pubsub.support.converter.SimplePubSubMessageConverter;
 import org.springframework.util.Assert;
 import org.springframework.util.concurrent.ListenableFuture;
-import org.springframework.util.concurrent.SettableListenableFuture;
 
 /**
  * Default implementation of {@link PubSubOperations}.
- *
  * <p>The main Google Cloud Pub/Sub integration component for publishing to topics and consuming
  * messages from subscriptions asynchronously or by pulling.
- *
  * @author Vinicius Carvalho
  * @author João André Martins
  * @author Mike Eltsufin
  * @author Chengyuan Zhao
+ * @author Doug Hoard
  */
 public class PubSubTemplate implements PubSubOperations, InitializingBean {
 
 	private static final Log LOGGER = LogFactory.getLog(PubSubTemplate.class);
 
-	private PubSubMessageConverter messageConverter = new SimplePubSubMessageConverter();
-
-	private final PublisherFactory publisherFactory;
-
-	private final SubscriberFactory subscriberFactory;
+	private final PubSubPublisherTemplate pubSubPublisherTemplate;
 
-	private final SubscriberStub subscriberStub;
-
-	private final PubSubAcknowledger acknowledger;
+	private final PubSubSubscriberTemplate pubSubSubscriberTemplate;
 
 	/**
-	 * Default {@link PubSubTemplate} constructor that uses {@link SimplePubSubMessageConverter}
-	 * to serialize and deserialize payloads.
+	 * Default {@link PubSubTemplate} constructor.
 	 * @param publisherFactory the {@link com.google.cloud.pubsub.v1.Publisher} factory to
-	 * publish to topics
+	 * publish to topics.
 	 * @param subscriberFactory the {@link com.google.cloud.pubsub.v1.Subscriber} factory
-	 * to subscribe to subscriptions
+	 * to subscribe to subscriptions.
 	 */
 	public PubSubTemplate(PublisherFactory publisherFactory,
-			SubscriberFactory subscriberFactory) {
-		this.publisherFactory = publisherFactory;
-		this.subscriberFactory = subscriberFactory;
-		this.subscriberStub = this.subscriberFactory.createSubscriberStub();
-		this.acknowledger = this.subscriberFactory.createAcknowledger();
+						SubscriberFactory subscriberFactory) {
+		Assert.notNull(publisherFactory, "A valid PublisherFactory is required.");
+		Assert.notNull(subscriberFactory, "A valid SubscriberFactory is required.");
+
+		this.pubSubPublisherTemplate = new PubSubPublisherTemplate(publisherFactory);
+		this.pubSubSubscriberTemplate = new PubSubSubscriberTemplate(subscriberFactory);
+	}
+
+	/**
+	 * Default {@link PubSubTemplate} constructor that uses a {@link PubSubPublisherTemplate}
+	 * and a {@link PubSubSubscriberTemplate}
+	 * @param pubSubPublisherTemplate the {@link PubSubPublisherTemplate} to
+	 * publish to topics.
+	 * @param pubSubSubscriberTemplate the {@link PubSubSubscriberTemplate} to
+	 * subscribe to subscriptions or pull messages
```

Comment:
This is a new `public` ctor in this class for the current new version therefore it must be marked with `@since` .